src/Traverser.php

```diff
@@ -2,198 +2,208 @@
+        'children' => 'children',
+    ];
+
     protected $query;
     protected $document;
-    protected $parentId;
-    protected $childrenIds;
-    protected $parentMethods;
-    protected $childrenMethods;
+    protected $relations = [];
+
+    public static function make(...$parameters)
+    {
+        return new static(...$parameters);
+    }
 
-    public function __construct($document)
+    public function __construct(Document $document, $relations = [])
     {
         $this->document = $document;
 
         $this->query = Query::eagerLoadAll();
 
-        $this->parentMethods = collect(
-            array_fill_keys(Prismic::$documents, 'parent')
-        );
-
-        $this->childrenMethods = collect(
-            array_fill_keys(Prismic::$documents, 'children')
-        );
+        $this->relations = collect($relations);
     }
 
-    public function setParentId($id): Traverser
-    {
-        $this->parentId = $id;
+    public function for($class, $parent = null, $children = null, $id = null) {
+        $this->relations->put($class, array_filter(compact('parent', 'children', 'id')));
 
         return $this;
     }
 
-    public function setChildrenIds($ids): Traverser
+    public function parentFor($class, $parent)
     {
-        $this->childrenIds = collect($ids);
-
-        return $this;
+        return $this->for($class, $parent, null, null);
     }
 
-    public function setParentMethods($methods): Traverser
+    public function childrenFor($class, $children)
     {
-        $this->parentMethods->merge($methods);
-
-        return $this;
+        return $this->for($class, null, $children, null);
     }
 
-    public function setChildrenMethods(array $methods): Traverser
+    public function idFor($class, $id)
     {
-        $this->childrenMethods->merge($methods);
-
-        return $this;
+        return $this->for($class, null, null, $id);
     }
 
-    public function parent(): ?Document
+    public function id()
     {
-        if ($this->parentId === false) {
-            return null;
-        }
-
-        if ($this->parentId) {
-            return $this->getParentFromId();
-        }
-
-        return $this->getParentFromChildren();
+        return $this->resolveRelation('id');
     }
 
-    protected function getParentFromId(): ?Document
+    public function parent(): ?Document
     {
-        return $this->query->find($this->parentId);
+        return $this->resolveRelation('parent');
     }
 
-    protected function getParentFromChildren(): ?Document
+    public function findParent($class = null): ?Document
     {
-        return $this->query->documentCache()->first(function ($document) {
-            $childrenMethod = $this->getChildrenMethod($document);
-
-            if (method_exists($document, $childrenMethod)) {
-                return $document->{$childrenMethod}()
-                    ->filter()
+        return $this->query->documentCache()
+            ->reject(function ($document) use ($class) {
```

Comment:
Remind me why we're passing in a class name?

Reply:
It's so you can find a parent of a particular type